bigquery/storage_iterator.go

```diff
@@ -0,0 +1,267 @@
+	decoder *arrowDecoder
+	records chan arrowRecordBatch
+
+	session *readSession
+}
+
+type arrowRecordBatch []byte
+
+func newStorageRowIteratorFromTable(ctx context.Context, table *Table) (*RowIterator, error) {
+	md, err := table.Metadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rs, err := table.c.rc.sessionForTable(ctx, table)
+	if err != nil {
+		return nil, err
+	}
+	it, err := newStorageRowIterator(ctx, rs, md.NumRows)
+	if err != nil {
+		return nil, err
+	}
+	it.arrowIterator.schema = md.Schema
+	return it, nil
+}
+
+func newStorageRowIteratorFromJob(ctx context.Context, job *Job, totalRows uint64) (*RowIterator, error) {
+	cfg, err := job.Config()
+	if err != nil {
+		return nil, err
+	}
+	qcfg := cfg.(*QueryConfig)
+	if qcfg.Dst == nil {
+		// TODO: script job ?
+		return nil, fmt.Errorf("nil job destination table")
+	}
+	return newStorageRowIteratorFromTable(ctx, qcfg.Dst)
+}
+
+func newRawStorageRowIterator(ctx context.Context, rs *readSession) (*arrowIterator, error) {
+	arrowIt := &arrowIterator{
+		ctx:     ctx,
+		session: rs,
+		records: make(chan arrowRecordBatch, rs.rc.maxWorkerCount+1),
+		errs:    make(chan error, rs.rc.maxWorkerCount+1),
+	}
+	if rs.bqSession == nil {
+		err := rs.start()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return arrowIt, nil
+}
+
+func newStorageRowIterator(ctx context.Context, rs *readSession, totalRows uint64) (*RowIterator, error) {
+	arrowIt, err := newRawStorageRowIterator(ctx, rs)
+	if err != nil {
+		return nil, err
+	}
+	it := &RowIterator{
+		ctx:           ctx,
+		arrowIterator: arrowIt,
+		TotalRows:     totalRows,
+		rows:          [][]Value{},
+	}
+	it.nextFunc = nextFuncForStorageIterator(it)
+	it.pageInfo = &iterator.PageInfo{
+		Token:   "",
+		MaxSize: int(totalRows),
+	}
+	return it, nil
+}
+
+func nextFuncForStorageIterator(it *RowIterator) func() error {
+	return func() error {
+		if len(it.rows) > 0 {
+			return nil
+		}
+		arrowIt := it.arrowIterator
+		record, err := arrowIt.next()
+		if err == iterator.Done {
+			if len(it.rows) == 0 {
+				return iterator.Done
+			}
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		rows, err := arrowIt.decoder.decodeArrowRecords(record)
+		if err != nil {
+			return err
+		}
+		it.rows = rows
+		return nil
+	}
+}
+
+func (it *arrowIterator) init() error {
+	if it.decoder != nil { // Already initialized
+		return nil
+	}
+
+	bqSession := it.session.bqSession
+	if bqSession == nil {
+		return errors.New("read session not initialized")
+	}
+
+	streams := bqSession.Streams
+	if len(streams) == 0 {
+		return iterator.Done
+	}
+
+	if it.schema == nil {
+		meta, err := it.session.table.Metadata(it.ctx)
+		if err != nil {
+			return err
+		}
+		it.schema = meta.Schema
+	}
+
+	decoder, err := newArrowDecoderFromSession(it.session, it.schema)
+	if err != nil {
+		return err
+	}
+	it.decoder = decoder
+
+	wg := sync.WaitGroup{}
+	sem := semaphore.NewWeighted(int64(it.session.rc.maxWorkerCount))
+	go func() {
```

Comment:
I think this may race?  wg.Wait() before the first wg.Add will proceed, since there's nothing outstanding?  You could potentially do something like wg.Add(len(streams)) before starting up this and your reader goroutines.

Reply:
good point, I changed it here.